lib/util.js

```diff
@@ -134,35 +134,52 @@ util.getDelocalizedHostname = function getDelocalizedHostname(hostname) {
  *
  * @param {string[]} versions - A list of version numbers
  * @param {string} mode - Either "major" or "minor", ("patch" will return the
+ * @param {object} meta - metadata for a given packages {semverRanges, latest, staticVersions}
  * same array of versions)
  */
-util.maxVersionPerMode = function maxVersionPerMode(versions, mode) {
+util.maxVersionPerMode = function maxVersionPerMode(versions, mode, meta) {
+  const semverRanges = meta.semverRanges.join(' || ')
+  const includeLatest = meta.latest
+  const { staticVersions } = meta
   // Remove versions that don't fit the semantic versioning convention
   versions = versions.filter((version) => {
     return /^\d+\.\d+\.\d+$/.test(version)
   })
 
-  // Creates an object where the key is the version number up to the mode
-  // specified, and the value is the highest version in that set.
-  const greatestVersionInMode = {}
-  versions.forEach((version) => {
-    // Get the version number up to the mode specified. (turns 1.0.1 into 1 if
-    // mode is major)
-    let versionMode = version
-    if (mode === 'major') {
-      versionMode = /^(\d+)/.exec(version)[1]
-    } else if (mode === 'minor') {
-      versionMode = /^(\d+\.\d+)/.exec(version)[1]
+  if (mode === 'patch') {
+    return versions
+  }
+
+  const versionRegex = mode === 'major' ? /^(\d+)/ : /^(\d+\.\d+)/
+  const greatestVersions = []
+  const latestVersion = versions[versions.length - 1]
+
+  const versionsByMajor = versions.reduce((previousValue, version) => {
+    const versionMode = versionRegex.exec(version)[1]
+    if (!previousValue.hasOwnProperty(versionMode)) {
+      previousValue[versionMode] = []
     }
 
-    // Keep track of the greatest version version in this mode
-    const greatest = greatestVersionInMode[versionMode]
-    if (!greatest || semver.gt(version, greatest)) {
-      greatestVersionInMode[versionMode] = version
+    if (staticVersions.includes(version)) {
+      greatestVersions.push(version)
     }
-  })
 
-  // Return just the greatest version numbers of each version "mode"
-  const greatestVersions = Object.values(greatestVersionInMode)
+    previousValue[versionMode].push(version)
+    return previousValue
+  }, {})
+
+  // put the latest first to avoid duplicates of the same version
+  // when looking for max within a semver range
+  if (includeLatest) {
+    greatestVersions.push(latestVersion)
+  }
+
+  for (const [, value] of Object.entries(versionsByMajor)) {
```

Comment:
Why not use `Object.values()`?

Reply:
good idea. I've leaned so heavy on lodash before I joined here I always forget the best ways to iterate over an object